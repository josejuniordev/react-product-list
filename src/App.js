import React, { useEffect, useState } from 'react';
import './App.scss';
import { BrowserRouter, Switch, Route, Link, withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { Suspense, lazy } from 'react';
import { fetchProducts } from './ducks/products';
import { fetchDepartments } from './ducks/departments';
import Navigation from './components/navigation/Navigation';
import Button from './components/button/Button';
import Navbar from './components/navbar/Navbar';
import { appPrefix } from './constants/settings';


const ProductsListPage = lazy(() => import('./pages/ProductsListPage'));
const ProductDetailsPage = lazy(() => import('./pages/ProductDetailsPage'));

function App(
  {
    callFetchProducts,
    callFetchDepartments,
  }
) {
  const selectedLink = window.location.pathname.replace('/', '') || 'home';
  const [initialized, setInitialized] = useState(false);
  const [isNavigationOpen, setIsNavigationOpenTo] = useState(false);
  const [history, setHistory] = useState({});

  useEffect(() => {
    if (!initialized) {
      initialize();
    }
  }, [initialized]);

  const initialize = () => {
    callFetchProducts();
    callFetchDepartments();
    setInitialized(true);
  };

  const openNavigation = () => {
    setIsNavigationOpenTo(true);
  };

  const closeNavigation = () => {
    setIsNavigationOpenTo(false);
  };

  const backButtonClickHandler = () => {
    if (Object.keys(history).length) {
      history.goBack();
    }
  };

  return (
    <main>
      <BrowserRouter basename="">
        <Navbar onToggleButtonClick={openNavigation} actionButtons={
          [
            <Button onClick={backButtonClickHandler} type='link' icon={<span>&#8249;</span>} label='Voltar' highlight />
          ]
        } />
        <section className={`${appPrefix}content`}>
          <Suspense
            fallback={ <p>carregando...</p> }
          >
            <Switch>
              <Route exact path={ `${ process.env.PUBLIC_URL }/` }
                     component={ () => <ProductsListPage historyHandler={setHistory} appInitialized={ initialized }/> }/>
              <Route exact path={ `${ process.env.PUBLIC_URL }/product-detail` }
                     component={ () => <ProductDetailsPage historyHandler={setHistory} appInitialized={ initialized }/> }/>
            </Switch>
          </Suspense>
        </section>
        <Navigation onClose={ closeNavigation } open={ isNavigationOpen }/>
      </BrowserRouter>
    </main>
  );
}

const connectedApp = connect(false,
  dispatch => {
    return {
      callFetchProducts() {
        dispatch(fetchProducts());
      },
      callFetchDepartments() {
        dispatch(fetchDepartments());
      }
    }
  }
)(
  App
);

export default connectedApp;
