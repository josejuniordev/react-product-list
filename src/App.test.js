import React from 'react';
import { mount, shallow } from 'enzyme';
import App from './App';
import store from './store';
import { Provider } from 'react-redux';

describe('App Component', () => {
  test('Should render correctly', () => {
    const rendered = shallow(
      <Provider store={store}>
        <App />
      </Provider>
    );

    expect(rendered).toMatchSnapshot();
  })
});
