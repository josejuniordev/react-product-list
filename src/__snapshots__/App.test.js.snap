// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Connect(App) />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Connect(App) />,
      "value": Object {
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "storeState": Object {
          "cart": Object {
            "data": Array [],
            "isLoading": false,
          },
          "departments": Object {
            "data": Array [],
            "isLoading": false,
          },
          "products": Object {
            "currentInView": null,
            "data": Array [],
            "isLoading": false,
            "productInDetailMode": null,
          },
        },
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": null,
        "_currentValue2": null,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Connect(App) />,
        "value": Object {
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "storeState": Object {
            "cart": Object {
              "data": Array [],
              "isLoading": false,
            },
            "departments": Object {
              "data": Array [],
              "isLoading": false,
            },
            "products": Object {
              "currentInView": null,
              "data": Array [],
              "isLoading": false,
              "productInDetailMode": null,
            },
          },
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": null,
          "_currentValue2": null,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
